### ISSUE TITLE
LTFB interactions with data buffer

### ISSUE THREAD

**Issue by samadejacobs** [LLNL/lbann#491]
Experiment with 64 LTFB models on JAG dataset is crashing with the following error:
```
*************************************************************************
 This lbann_exception is about to be thrown:Error in calculating the partial mini-batch size, exceeds the max mini-batch size

 Am now attempting to print the stack trace ...
**************************************************************************
rank: 0 :: rank: 0 :: rank: 0 :: lbann::lbann_exception::lbann_exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
rank: 0 :: lbann::partitioned_io_buffer::calculate_num_iterations_per_epoch_spanning_models(int, lbann::generic_data_reader*)
rank: 0 :: lbann::generic_input_layer::setup_data()
```

For this case,  max_minibatch and last_mini_batch are computed as 64 and 82 respectively (it is not immediately clear how these values are computed). Broadly speaking, we need to investigate how data_reader and data_buffer interacts in LTFB multimodel setting. I am concern that global minibatch is still being reported which I think might be affecting how number of iterations are computed. For LTFB, global minibatch (and its use for other computations e.g. number of iterations) doesn't make much sense because the models are "independent".

**Comment by bvanessen** [comment]
I will look into this, but the last mini-batch should always be smaller or equal to the normal mini-batch.  Also, global mini-batch should be set to the model's mini-batch size for LTFB, thus having no effect.

**Comment by samadejacobs** [comment]
Yes you are right, global mini batch should be set to model's mini batch for LTFB (or independent multi-models broadly), I don't think that is the case right now and is possibly having some unintended impacts on LTFB results.

**Comment by samadejacobs** [comment]
By setting m_comm->get_num_models() to 1 in line number 301 in [distributed_io_buffer.cpp](https://github.com/LLNL/lbann/blob/develop/src/io/data_buffers/distributed_io_buffer.cpp), the reported crash was fixed and so is consistency in reporting same local mini-batch and global mini-batch and same training number of iterations for LTFB multi-models as number of models varies e.g.
(a) 4 LTFB models (each model at 64 mini-batch size)
```
--------------------------------------------------------------------------------
[0] Epoch : stats formated [tr/v/te] iter/epoch = [1125/125/1]
            global MB = [  64/  64/  16] global last MB = [  46  /  62  /  16  ]
             local MB = [  64/  64/  16]  local last MB = [  46+0/  62+0/  16+0]
--------------------------------------------------------------------------------
```
(b) 16 LTFB models (each model at 64 mini-batch size)
```
--------------------------------------------------------------------------------
[0] Epoch : stats formated [tr/v/te] iter/epoch = [1125/125/1]
            global MB = [  64/  64/  16] global last MB = [  46  /  62  /  16  ]
             local MB = [  64/  64/  16]  local last MB = [  46+0/  62+0/  16+0]
--------------------------------------------------------------------------------
```
Prior to this hack, reported stats vary with the number of models which is not what we want (this impacts on using correct iterations/round size for tournament):
(c) 4 LTFB models (each model at 64 mini-batch size)
```
--------------------------------------------------------------------------------
[0] Epoch : stats formated [tr/v/te] iter/epoch = [282/125/1]
            global MB = [ 256/  64/  16] global last MB = [  46  /  62  /  16  ]
             local MB = [  64/  64/  16]  local last MB = [  13+0/  62+0/  16+0]
--------------------------------------------------------------------------------
```
(d) 16 LTFB models (each model at 64 mini-batch size)
```
--------------------------------------------------------------------------------
[0] Epoch : stats formated [tr/v/te] iter/epoch = [71/125/1]
            global MB = [1024/  64/  16] global last MB = [ 302  /  62  /  16  ]
             local MB = [  64/  64/  16]  local last MB = [  32+0/  62+0/  16+0]
--------------------------------------------------------------------------------
```
However, I am afraid that this hack might have introduced other unintentional bugs in how data is ingested/processed , so Brian is in the best position to know/investigate.

**Comment by samadejacobs** [comment]
An easy (appropriate) fix for this issue is to set the data_set_per_model in Input layer to true, it is false by default. I made this change in prototext and things seem to be working. I will close this issue as soon am able to confirm no other closely related issue.

**Comment by davidHysom** [comment]
This should perhaps be a different issue -- ??

I was getting errors for an allocated data set (aka, partitioned dataset) when using a distributed io_buffer:


  include/lbann/layers/io/input/generic_input_layer.hpp 278 :: I/O layers number of samples processed (0) does not match the mini-batch size (2)


Per Sam's comment, I set the data_set_per_model in Input layer to true, and the exception is no longer being thrown. For debugging I have each "train" data_reader print to file the indices it uses in calls to data_reader::fetch_data(). For three models with three processors per model, and using the mnist data set, mb_size is 64, except for the final call for procs 3, 5, and 8, in which case it's 4.

So I believe that this is working properly.

________________________________

**Comment by davidHysom** [comment]
When I started writing my previous I thought there was a problem (issue). But it looks like everything's working, so it's a non-issue.

________________________________
